core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -861,27 +861,13 @@ private[sql] final class PreprocessTableInsertOrPut(conf: SQLConf)
     case d@DeleteFromTable(table, child) if table.resolved && child.resolved =>
       EliminateSubqueryAliases(table) match {
         case l@LogicalRelation(dr: DeletableRelation, _, _) =>
-          def comp(a: Attribute, targetCol: String): Boolean = a match {
-            case ref: AttributeReference => targetCol.equals(ref.name.toUpperCase)
-          }
-          // First, make sure the where column(s) of the delete are in schema of the relation.
           val expectedOutput = l.output
-          if (!child.output.forall(a => expectedOutput.exists(e => comp(a, e.name.toUpperCase)))) {
+          if (child.output.length != expectedOutput.length) {
             throw new AnalysisException(s"$l requires that the query in the " +
                 "WHERE clause of the DELETE FROM statement " +
-                "generates the same column name(s) as in its schema but found " +
+                "generates the same number of column(s) as in its schema but found " +
                 s"${child.output.mkString(",")} instead.")
           }
-          l match {
-            case LogicalRelation(ps: PartitionedDataSourceScan, _, _) =>
-              if (!ps.partitionColumns.forall(a => child.output.exists(e =>
-                comp(e, a.toUpperCase)))) {
-                throw new AnalysisException(s"${child.output.mkString(",")}" +
-                    s" columns in the WHERE clause of the DELETE FROM statement must " +
-                    s"have all the parititioning column(s) ${ps.partitionColumns.mkString(",")}.")
-              }
-            case _ =>
-          }
```

Comment:
I think this is trying to enforce that partitioning columns should be part of delete, at least for row tables. This is a good thing to enforce else the deleteFrom will be very inefficient -- every row delete on row table will lead to a full scan in store side which will be very expensive. I think the problem in SNAP-2157 is that no partitioning has been defined for the table (try if that works by adding a partition_by C_CUSTKEY option). If that's the case perhaps the fix is to change default partitioning to be on primary key columns, if defined, for row tables.

Reply:
DeleteFromTable only works if all the columns of the table match. Although the column names might not match. Hence it will automatically include column tables. But yes that validation also got skipped in this change for MutableRelations. Adding them back.

Reply:
I am still not able to pinpoint the exact cause of SNAP-2157, as to why it was only failing for decimal types. When a primary key is defined it takes the join key as taken as PK. If partition key is there PK + partition keys are considered.